src/github.com/getlantern/balancer/balancer.go

```diff
@@ -95,20 +75,19 @@ func (b *Balancer) dialerAndConn(network, addr string, targetQOS int) (*Dialer,
 	// To prevent dialing infinitely
 	attempts := 3
 	for i := 0; i < attempts; i++ {
-		if len(dialers) == 0 {
+		if dialers.Len() == 0 {
 			return nil, nil, fmt.Errorf("No dialers left to try on pass %v", i)
 		}
 		var d *dialer
-		d, dialers = randomDialer(dialers, targetQOS)
+		d = heap.Pop(&dialers).(*dialer)
```

Comment:
Also, what if there are multiple dialers with effectively the same priority (or similar priorities)?